File: typescript/src/protocol.ts
/**
 * The manka-shreds wire protocol: frames, control messages and message layouts.
 *
 * Everything here is byte layout. It is kept separate from the client so that a consumer wanting
 * to drive the protocol some other way — a browser transport, a different runtime, a proxy — can
 * use the parsing without the socket handling.
 *
 * Offsets are fixed and every section is 8-byte aligned, which is why fields are read by pointing
 * at them rather than by walking length prefixes.
 */

/** Protocol version this implementation speaks. */
export const PROTOCOL_VERSION = 1;

/** Bytes in a frame header. */
export const FRAME_HEADER_LEN = 16;

/** Largest payload a frame may carry. */
export const MAX_FRAME_PAYLOAD = 16 << 20;

/** What a frame carries. */
export enum FrameKind {
  /** A kind this build does not recognise. Never sent; produced when reading. */
  Unknown = 0,
  Transaction = 1,
  Entry = 2,
  SlotStart = 3,
  SlotEnd = 4,
  Duplicate = 5,
  RawShred = 6,
  Lag = 7,
  SourceStats = 8,
  Hello = 9,
  HelloAck = 10,
  FilterAck = 11,
  Error = 12,
  Ping = 13,
  Pong = 14,
  SetFilter = 15,
  /** The server's nonce and its proof that it holds the key. */
  Challenge = 16,
  /** The client's proof that it holds the key. */
  Prove = 17,
  /** The compression dictionary the server wants this connection to use. */
  Dictionary = 18,
}

/** Streams a subscriber may ask for. A key grants a subset; the server grants the intersection. */
export const Stream = {
  TRANSACTIONS: 1 << 0,
  ENTRIES: 1 << 1,
  SLOT_EVENTS: 1 << 2,
  DUPLICATES: 1 << 3,
  /**
   * Every shred verbatim, before the node verifies or decodes anything.
   *
   * The highest-rate stream published — a multiple of `TRANSACTIONS`, since a shred is a fragment
   * and most fragments carry votes. Nothing on it is claimed to be leader-signed.
   */
  RAW_SHREDS: 1 << 4,
  /** Reserved on the wire; no producer exists and the name is refused at key load. */
  SOURCE_STATS: 1 << 5,
  VOTES: 1 << 6,
} as const;

/** Every stream that is actually published. */
export const ALL_STREAMS =
  Stream.TRANSACTIONS |
  Stream.ENTRIES |
  Stream.SLOT_EVENTS |
  Stream.DUPLICATES |
  Stream.RAW_SHREDS |
  Stream.VOTES;

/** Compression codecs. The value is what travels in the frame flags. */
export enum Codec {
  None = 0,
  Lz4 = 1,
  Zstd = 2,
}

/** Bits offered in a handshake, one per codec this client can decode. */
export const CodecBit = {
  NONE: 1 << 0,
  LZ4: 1 << 1,
  ZSTD: 1 << 2,
} as const;

/**
 * What a client can accept beyond the streams it subscribes to.
 *
 * A bitmask in the greeting, in a byte that was previously padding. Absent bits mean "no", which is
 * what a client built before this existed should be taken to have said.
 */
export const Capability = {
  /**
   * This client will accept the server's compression dictionary on this connection.
   *
   * Without it the server never sends one. With it, a client holding no dictionary — or a different
   * one — is brought up to the server's during the handshake and streams at the full ratio from its
   * first message, rather than paying roughly three times the bandwidth until somebody notices it
   * was never given one.
   */
  ACCEPTS_DICTIONARY: 1 << 0,
} as const;

/**
 * Largest dictionary this client will accept from a server.
 *
 * The one thing a server hands a client that is neither fixed-width nor bounded by something the
 * client asked for, so it needs a bound of its own — otherwise a hostile or compromised node
 * streams as much "dictionary" as this side will allocate. Far past any dictionary worth training:
 * the ratio stops improving well below it.
 */
export const MAX_DICTIONARY_BYTES = 8 << 20;

/**
 * Largest control message this client will read while the handshake is in progress.
 *
 * A challenge is sixty-four bytes and an acknowledgement is smaller; the only variable one is an
 * error, whose detail is a sentence. It needs a bound because the length is the *server's* to
 * choose and it is what this side buffers against — and at this point in the exchange the server
 * has proved nothing. Judged by {@link MAX_FRAME_PAYLOAD} instead, any address you dial could hold
 * sixteen megabytes of yours for the price of a header.
 *
 * The dictionary is the one handshake frame this does not apply to; it has
 * {@link MAX_DICTIONARY_BYTES} of its own, and it arrives only after the server has proved itself.
 */
export const MAX_HANDSHAKE_BYTES = 64 * 1024;

/** The compression dictionary a connection will use, as the server sends it. */
export interface Dictionary {
  /** Identifier of these bytes, matching what the acknowledgement announced. */
  id: number;
  /** The dictionary itself. */
  bytes: Buffer;
}

/** Bytes before a dictionary's payload. */
export const DICTIONARY_HEADER_LEN = 8;

/**
 * Reads a dictionary message, refusing one larger than {@link MAX_DICTIONARY_BYTES}.
 *
 * The declared length is checked *before* the body is taken, so a number a peer invented cannot
 * decide how much this side allocates.
 */
export function readDictionary(buf: Buffer): Dictionary {
  if (buf.length < DICTIONARY_HEADER_LEN) {
    throw new ProtocolError(
      `dictionary needs ${DICTIONARY_HEADER_LEN} bytes, have ${buf.length}`,
    );
  }
  const id = buf.readUInt32LE(0);
  const len = buf.readUInt32LE(4);
  if (len > MAX_DICTIONARY_BYTES) {
    throw new ProtocolError(
      `the server sent a ${len}-byte dictionary; the limit is ${MAX_DICTIONARY_BYTES}`,
    );
  }
  if (buf.length < DICTIONARY_HEADER_LEN + len) {
    throw new ProtocolError(
      `dictionary needs ${DICTIONARY_HEADER_LEN + len} bytes, have ${buf.length}`,
    );
  }
  return {
    id,
    // Copied rather than sliced: a subarray keeps the whole read buffer alive, and this one is
    // held for the life of the connection.
    bytes: Buffer.from(buf.subarray(DICTIONARY_HEADER_LEN, DICTIONARY_HEADER_LEN + len)),
  };
}

/**
 * Encodes a dictionary message.
 *
 * A client never sends one. This exists so a test can build what a server would.
 */
export function writeDictionary(dictionary: Dictionary): Buffer {
  const out = Buffer.allocUnsafe(DICTIONARY_HEADER_LEN + dictionary.bytes.length);
  out.writeUInt32LE(dictionary.id, 0);
  out.writeUInt32LE(dictionary.bytes.length, 4);
  dictionary.bytes.copy(out, DICTIONARY_HEADER_LEN);
  return out;
}

/** Why the server refused or closed. */
export enum ErrorCode {
  Unauthorized = 1,
  Forbidden = 2,
  TooSlow = 3,
  RateLimited = 4,
  BadRequest = 5,
  Shutdown = 6,
  UnsupportedVersion = 7,
}

/** How much a transaction's FEC set could be vouched for. */
export enum Verification {
  /** The leader's signature over the set's merkle root checked out. */
  Verified = 'verified',
  /** The node runs with verification switched off. */
  Disabled = 'disabled',
  /** No leader is known for the slot. */
  UnknownLeader = 'unknown-leader',
  /** The leader schedule is behind. */
  StaleSchedule = 'stale-schedule',
}

/** A decoded frame header. */
export interface FrameHeader {
  /** Payload bytes following the header. */
  len: number;
  /** What the payload is, as far as this build recognises it. */
  kind: FrameKind;
  /** The kind byte exactly as it arrived, so an unknown frame keeps its identity. */
  rawKind: number;
  /** Whether the payload is compressed. */
  compressed: boolean;
  /** Codec the payload was compressed with. */
  codec: Codec;
  /**
   * Which of this connection's named filters the frame matched, as a bitmask.
   *
   * Zero when the connection named no filters. Bit `i` is the filter at index `i` in the set that
   * was submitted; one transaction can match several and is delivered once with every match
   * recorded, rather than once per filter.
   */
  matched: number;
  /** Per-connection sequence number. A gap is exactly what was dropped. */
  seq: bigint;
}

/**
 * Reads a frame header.
 *
 * An unrecognised kind is not an error: a client built before a frame type existed must skip it,
 * not drop the connection, or adding one would break every deployed subscriber.
 */
export function readFrameHeader(buf: Buffer, at = 0): FrameHeader {
  if (buf.length - at < FRAME_HEADER_LEN) {
    throw new ProtocolError(`frame header needs ${FRAME_HEADER_LEN} bytes, have ${buf.length - at}`);
  }
  const len = buf.readUInt32LE(at);
  if (len > MAX_FRAME_PAYLOAD) {
    throw new ProtocolError(`frame declares ${len} payload bytes, limit is ${MAX_FRAME_PAYLOAD}`);
  }
  const rawKind = buf.readUInt8(at + 4);
  const flags = buf.readUInt8(at + 5);
  return {
    len,
    kind: rawKind in FrameKind ? (rawKind as FrameKind) : FrameKind.Unknown,
    rawKind,
    compressed: (flags & 0b100) !== 0,
    codec: (flags & 0b11) as Codec,
    matched: buf.readUInt16LE(at + 6),
    seq: buf.readBigUInt64LE(at + 8),
  };
}

/** Writes a frame with `payload` as its body. */
export function writeFrame(
  kind: FrameKind,
  seq: bigint,
  payload: Buffer,
  matched = 0,
): Buffer {
  const out = Buffer.allocUnsafe(FRAME_HEADER_LEN + payload.length);
  out.writeUInt32LE(payload.length, 0);
  out.writeUInt8(kind, 4);
  out.writeUInt8(0, 5);
  out.writeUInt16LE(matched, 6);
  out.writeBigUInt64LE(seq, 8);
  payload.copy(out, FRAME_HEADER_LEN);
  return out;
}

/** Anything the protocol could not be read as. */
export class ProtocolError extends Error {
  constructor(message: string) {
    super(message);
    this.name = 'ProtocolError';
  }
}

/** The server refused the connection, or closed it. */
export class ServerError extends Error {
  constructor(
    readonly code: ErrorCode,
    readonly detail: string,
  ) {
    super(`server error ${ErrorCode[code] ?? code}: ${detail}`);
    this.name = 'ServerError';
  }
}

// ---------------------------------------------------------------------------------------------
// Handshake
// ---------------------------------------------------------------------------------------------

/**
 * What a client offers when it connects.
 *
 * **Carries no credential.** Your key is named by a one-way reference and proved afterwards — see
 * `handshake`. A captured greeting is worth nothing.
 */
export interface Hello {
  streams: number;
  codecs: number;
  /** Which key is claimed, without disclosing it. 32 bytes. */
  keyRef: Buffer;
  /** Fresh per connection, so a captured server proof cannot be replayed at this client. 32 bytes. */
  clientNonce: Buffer;
  /**
   * What this client can be sent beyond the streams themselves — see {@link Capability}.
   *
   * Occupies a byte the greeting has always written as zero and never read, so a client built
   * before this existed advertises nothing, which is the correct reading of its silence.
   */
  capabilities: number;
  /** Content hash of the dictionary this client holds, or zero for none. */
  dictionaryId: number;
  /** A filter document to install before the stream starts, if any. */
  filter?: string;
}

/**
 * Encodes a greeting.
 *
 * The reference and the nonce are fixed width, so the greeting has no variable-length credential
 * fields — there is no credential in it to be variable. The dictionary id trails them so a peer that
 * stops reading there still parses: a missing id simply means no dictionary. The filter trails that
 * in turn, length-prefixed with a `u32` because a filter naming two thousand accounts is far past
 * 64 KiB.
 */
export function writeHello(hello: Hello): Buffer {
  const filter = hello.filter === undefined ? undefined : Buffer.from(hello.filter, 'utf8');
  const base = 8 + 32 + 32 + 4;
  const out = Buffer.allocUnsafe(base + (filter === undefined ? 0 : 4 + filter.length));
  out.writeUInt16LE(PROTOCOL_VERSION, 0);
  out.writeUInt32LE(hello.streams, 2);
  out.writeUInt8(hello.codecs, 6);
  out.writeUInt8(hello.capabilities, 7);
  hello.keyRef.copy(out, 8);
  hello.clientNonce.copy(out, 40);
  out.writeUInt32LE(hello.dictionaryId, 72);
  if (filter !== undefined) {
    // Byte length, not character count: a filter is base58 today but the field is UTF-8, and
    // `string.length` would understate any key that is not.
    out.writeUInt32LE(filter.length, base);
    filter.copy(out, base + 4);
  }
  return out;
}

/**
 * The server's nonce and its proof that it holds your key.
 *
 * Sent before you have proved anything. Checking it is what tells you the peer is the node rather
 * than something sitting in front of it.
 */
export interface Challenge {
  serverNonce: Buffer;
  proof: Buffer;
}

/** Decodes a challenge. */
export function readChallenge(buf: Buffer): Challenge {
  if (buf.length < 64) throw new ProtocolError('challenge truncated');
  return {
    serverNonce: Buffer.from(buf.subarray(0, 32)),
    proof: Buffer.from(buf.subarray(32, 64)),
  };
}

/** Encodes your proof that you hold the key. */
export function writeProve(proof: Buffer): Buffer {
  return Buffer.from(proof);
}

/** What the server settled. */
export interface HelloAck {
  protocolVersion: number;
  sessionId: bigint;
  /** Streams actually granted — the intersection of what was asked and what the key permits. */
  granted: number;
  codec: Codec;
  /** How often the server pings an idle connection. */
  keepaliveMs: number;
  /** Non-zero only when the server holds the same dictionary this client offered. */
  dictionaryId: number;
}

/** Bytes in an encoded acknowledgement. */
export const HELLO_ACK_LEN = 22;

/** Decodes an acknowledgement. */
export function readHelloAck(buf: Buffer): HelloAck {
  if (buf.length < HELLO_ACK_LEN) {
    throw new ProtocolError(`hello ack needs ${HELLO_ACK_LEN} bytes, have ${buf.length}`);
  }
  return {
    protocolVersion: buf.readUInt16LE(0),
    granted: buf.readUInt32LE(2),
    sessionId: buf.readBigUInt64LE(6),
    codec: buf.readUInt8(14) as Codec,
    // Byte 15 is reserved.
    keepaliveMs: buf.readUInt16LE(16),
    dictionaryId: buf.readUInt32LE(18),
  };
}

/** Decodes an error message. */
export function readError(buf: Buffer): ServerError {
  if (buf.length < 4) {
    throw new ProtocolError('error message truncated');
  }
  const code = buf.readUInt16LE(0);
  const len = buf.readUInt16LE(2);
  const detail = buf.subarray(4, 4 + len).toString('utf8');
  return new ServerError(code as ErrorCode, detail);
}

/** Frames this client missed, and where the stream resumes. */
export interface Lag {
  dropped: bigint;
  resumeSeq: bigint;
}

/** Decodes a lag report. */
export function readLag(buf: Buffer): Lag {
  if (buf.length < 16) throw new ProtocolError('lag message truncated');
  return { dropped: buf.readBigUInt64LE(0), resumeSeq: buf.readBigUInt64LE(8) };
}

/** The server's verdict on a submitted filter. */
export interface FilterAck {
  accepted: boolean;
  cost: number;
  detail: string;
}

/** Decodes a filter acknowledgement. */
export function readFilterAck(buf: Buffer): FilterAck {
  if (buf.length < 8) throw new ProtocolError('filter ack truncated');
  const len = buf.readUInt16LE(6);
  return {
    accepted: buf.readUInt8(0) !== 0,
    // Byte 1 is reserved.
    cost: buf.readUInt32LE(2),
    detail: buf.subarray(8, 8 + len).toString('utf8'),
  };
}

/** One filter, with the name the subscriber chose for it. */
export interface NamedFilter {
  /** What you call it. Comes back in a refusal so you know which one was at fault. */
  name?: string;
  /** The filter itself. */
  spec: object;
}

/**
 * Encodes a filter submission. The frame length delimits it, so the JSON needs no prefix.
 *
 * The whole set replaces whatever the connection carried, and is accepted or refused together — a
 * connection is never left holding part of what was asked for.
 */
export function writeSetFilter(filters: NamedFilter[]): Buffer {
  return Buffer.from(JSON.stringify({ filters }), 'utf8');
}

/**
 * The content hash a dictionary is identified by.
 *
 * FNV-1a folded to 32 bits, matching the server exactly. Two peers agree on a dictionary by
 * agreeing on this number, so it must be computed the same way on both sides or the dictionary is
 * silently never used.
 */
export function dictionaryId(dictionary: Buffer): number {
  if (dictionary.length === 0) return 0;
  let hash = 0xcbf29ce484222325n;
  const prime = 0x100000001b3n;
  const mask = 0xffffffffffffffffn;
  for (const byte of dictionary) {
    hash = ((hash ^ BigInt(byte)) * prime) & mask;
  }
  const folded = Number((hash ^ (hash >> 32n)) & 0xffffffffn);
  // Zero means "no dictionary", so a hash that lands there is nudged rather than misread.
  return folded === 0 ? 1 : folded;
}

File: python/src/manka_shreds_sdk/protocol.py
"""The manka-shreds wire protocol: frames, control messages and message layouts.

Everything here is byte layout. It is kept separate from the client so that a consumer wanting to
drive the protocol some other way — a different transport, a proxy, a replay tool — can use the
parsing without the socket handling.

Offsets are fixed and every section is 8-byte aligned, which is why fields are read by pointing at
them rather than by walking length prefixes.
"""

from __future__ import annotations

import json
import struct
from dataclasses import dataclass
from enum import IntEnum, StrEnum

__all__ = [
    "ALL_STREAMS",
    "Capability",
    "Challenge",
    "Codec",
    "CodecBit",
    "DICTIONARY_HEADER_LEN",
    "Dictionary",
    "ErrorCode",
    "FRAME_HEADER_LEN",
    "FilterAck",
    "FrameHeader",
    "FrameKind",
    "HELLO_ACK_LEN",
    "Hello",
    "HelloAck",
    "Lag",
    "MAX_DICTIONARY_BYTES",
    "MAX_FRAME_PAYLOAD",
    "MAX_HANDSHAKE_BYTES",
    "NamedFilter",
    "PROTOCOL_VERSION",
    "ProtocolError",
    "ServerError",
    "Stream",
    "Verification",
    "dictionary_id",
    "read_challenge",
    "read_dictionary",
    "read_error",
    "read_filter_ack",
    "read_frame_header",
    "read_hello_ack",
    "read_lag",
    "write_dictionary",
    "write_frame",
    "write_hello",
    "write_prove",
    "write_set_filter",
]

#: Protocol version this implementation speaks.
PROTOCOL_VERSION = 1

#: Bytes in a frame header.
FRAME_HEADER_LEN = 16

#: Largest payload a frame may carry.
MAX_FRAME_PAYLOAD = 16 << 20


class FrameKind(IntEnum):
    """What a frame carries."""

    #: A kind this build does not recognise. Never sent; produced when reading.
    UNKNOWN = 0
    TRANSACTION = 1
    ENTRY = 2
    SLOT_START = 3
    SLOT_END = 4
    DUPLICATE = 5
    RAW_SHRED = 6
    LAG = 7
    SOURCE_STATS = 8
    HELLO = 9
    HELLO_ACK = 10
    FILTER_ACK = 11
    ERROR = 12
    PING = 13
    PONG = 14
    SET_FILTER = 15
    #: The server's nonce and its proof that it holds the key.
    CHALLENGE = 16
    #: The client's proof that it holds the key.
    PROVE = 17
    #: The compression dictionary the server wants this connection to use.
    DICTIONARY = 18


class Stream(IntEnum):
    """Streams a subscriber may ask for.

    A key grants a subset; the server grants the intersection.
    """

    TRANSACTIONS = 1 << 0
    ENTRIES = 1 << 1
    SLOT_EVENTS = 1 << 2
    DUPLICATES = 1 << 3
    #: Every shred verbatim, before the node verifies or decodes anything.
    #:
    #: The highest-rate stream published — a multiple of :attr:`TRANSACTIONS`, since a shred is a
    #: fragment and most fragments carry votes. Nothing on it is claimed to be leader-signed.
    RAW_SHREDS = 1 << 4
    #: Reserved on the wire; no producer exists and the name is refused at key load.
    SOURCE_STATS = 1 << 5
    VOTES = 1 << 6


#: Every stream that is actually published.
ALL_STREAMS = (
    Stream.TRANSACTIONS
    | Stream.ENTRIES
    | Stream.SLOT_EVENTS
    | Stream.DUPLICATES
    | Stream.RAW_SHREDS
    | Stream.VOTES
)


class Codec(IntEnum):
    """Compression codecs. The value is what travels in the frame flags."""

    NONE = 0
    LZ4 = 1
    ZSTD = 2


class CodecBit(IntEnum):
    """Bits offered in a handshake, one per codec this client can decode."""

    NONE = 1 << 0
    LZ4 = 1 << 1
    ZSTD = 1 << 2


class Capability(IntEnum):
    """What a client can accept beyond the streams it subscribes to.

    A bitmask in the greeting, in a byte that was previously padding. Absent bits mean "no", which
    is what a client built before this existed should be taken to have said.
    """

    #: This client will accept the server's compression dictionary on this connection.
    #:
    #: Without it the server never sends one. With it, a client holding no dictionary — or a
    #: different one — is brought up to the server's during the handshake and streams at the full
    #: ratio from its first message, rather than paying roughly three times the bandwidth until
    #: somebody notices it was never given one.
    ACCEPTS_DICTIONARY = 1 << 0


#: Largest dictionary this client will accept from a server.
#:
#: The one thing a server hands a client that is neither fixed-width nor bounded by something the
#: client asked for, so it needs a bound of its own — otherwise a hostile or compromised node
#: streams as much "dictionary" as this side will allocate.
MAX_DICTIONARY_BYTES = 8 << 20

#: Largest control message this client will read while the handshake is in progress.
#:
#: A challenge is sixty-four bytes and an acknowledgement is smaller; the only variable one is an
#: error, whose detail is a sentence. It needs a bound because the length is the *server's* to
#: choose and it is what this side buffers against — and at this point in the exchange the server
#: has proved nothing.
MAX_HANDSHAKE_BYTES = 64 * 1024


class ProtocolError(Exception):
    """Anything the protocol could not be read as."""


class ServerError(Exception):
    """The server refused the connection, or closed it."""

    def __init__(self, code: int, detail: str) -> None:
        try:
            name = ErrorCode(code).name
        except ValueError:
            name = str(code)
        super().__init__(f"server error {name}: {detail}")
        self.code = code
        self.detail = detail


class ErrorCode(IntEnum):
    """Why the server refused or closed."""

    UNAUTHORIZED = 1
    FORBIDDEN = 2
    TOO_SLOW = 3
    RATE_LIMITED = 4
    BAD_REQUEST = 5
    SHUTDOWN = 6
    UNSUPPORTED_VERSION = 7


class Verification(StrEnum):
    """How much a transaction's FEC set could be vouched for."""

    #: The leader's signature over the set's merkle root checked out.
    VERIFIED = "verified"
    #: The node runs with verification switched off.
    DISABLED = "disabled"
    #: No leader is known for the slot.
    UNKNOWN_LEADER = "unknown-leader"
    #: The leader schedule is behind.
    STALE_SCHEDULE = "stale-schedule"


@dataclass(frozen=True, slots=True)
class Dictionary:
    """The compression dictionary a connection will use, as the server sends it."""

    #: Identifier of these bytes, matching what the acknowledgement announced.
    id: int
    #: The dictionary itself.
    bytes: bytes


#: Bytes before a dictionary's payload.
DICTIONARY_HEADER_LEN = 8


def read_dictionary(buf: bytes) -> Dictionary:
    """Read a dictionary message, refusing one larger than :data:`MAX_DICTIONARY_BYTES`.

    The declared length is checked *before* the body is taken, so a number a peer invented cannot
    decide how much this side allocates.
    """
    if len(buf) < DICTIONARY_HEADER_LEN:
        raise ProtocolError(f"dictionary needs {DICTIONARY_HEADER_LEN} bytes, have {len(buf)}")
    ident, length = struct.unpack_from("<II", buf, 0)
    if length > MAX_DICTIONARY_BYTES:
        raise ProtocolError(
            f"the server sent a {length}-byte dictionary; the limit is {MAX_DICTIONARY_BYTES}"
        )
    if len(buf) < DICTIONARY_HEADER_LEN + length:
        raise ProtocolError(
            f"dictionary needs {DICTIONARY_HEADER_LEN + length} bytes, have {len(buf)}"
        )
    return Dictionary(
        id=ident,
        bytes=bytes(buf[DICTIONARY_HEADER_LEN : DICTIONARY_HEADER_LEN + length]),
    )


def write_dictionary(dictionary: Dictionary) -> bytes:
    """Encode a dictionary message.

    A client never sends one. This exists so a test can build what a server would.
    """
    return struct.pack("<II", dictionary.id, len(dictionary.bytes)) + dictionary.bytes


@dataclass(frozen=True, slots=True)
class FrameHeader:
    """A decoded frame header."""

    #: Payload bytes following the header.
    len: int
    #: What the payload is, as far as this build recognises it.
    kind: FrameKind
    #: The kind byte exactly as it arrived, so an unknown frame keeps its identity.
    raw_kind: int
    #: Whether the payload is compressed.
    compressed: bool
    #: Codec the payload was compressed with.
    codec: Codec
    #: Which of this connection's named filters the frame matched, as a bitmask.
    #:
    #: Zero when the connection named no filters. Bit ``i`` is the filter at index ``i`` in the set
    #: that was submitted; one transaction can match several and is delivered once with every match
    #: recorded, rather than once per filter.
    matched: int
    #: Per-connection sequence number. A gap is exactly what was dropped.
    seq: int


def read_frame_header(buf: bytes, at: int = 0) -> FrameHeader:
    """Read a frame header.

    An unrecognised kind is not an error: a client built before a frame type existed must skip it,
    not drop the connection, or adding one would break every deployed subscriber.
    """
    if len(buf) - at < FRAME_HEADER_LEN:
        raise ProtocolError(f"frame header needs {FRAME_HEADER_LEN} bytes, have {len(buf) - at}")
    length, raw_kind, flags, matched, seq = struct.unpack_from("<IBBHQ", buf, at)
    if length > MAX_FRAME_PAYLOAD:
        raise ProtocolError(f"frame declares {length} payload bytes, limit is {MAX_FRAME_PAYLOAD}")
    try:
        kind = FrameKind(raw_kind)
    except ValueError:
        kind = FrameKind.UNKNOWN
    return FrameHeader(
        len=length,
        kind=kind,
        raw_kind=raw_kind,
        compressed=(flags & 0b100) != 0,
        codec=Codec(flags & 0b11),
        matched=matched,
        seq=seq,
    )


def write_frame(kind: FrameKind, seq: int, payload: bytes, matched: int = 0) -> bytes:
    """Write a frame with ``payload`` as its body."""
    return struct.pack("<IBBHQ", len(payload), int(kind), 0, matched, seq) + payload


# -------------------------------------------------------------------------------------------------
# Handshake
# -------------------------------------------------------------------------------------------------


@dataclass(frozen=True, slots=True)
class Hello:
    """What a client offers when it connects.

    **Carries no credential.** Your key is named by a one-way reference and proved afterwards — see
    :mod:`manka_shreds_sdk.handshake`. A captured greeting is worth nothing.
    """

    streams: int
    codecs: int
    #: Which key is claimed, without disclosing it. 32 bytes.
    key_ref: bytes
    #: Fresh per connection, so a captured server proof cannot be replayed at this client. 32 bytes.
    client_nonce: bytes
    #: What this client can be sent beyond the streams themselves — see :class:`Capability`.
    capabilities: int = 0
    #: Content hash of the dictionary this client holds, or zero for none.
    dictionary_id: int = 0
    #: A filter document to install before the stream starts, if any.
    filter: str | None = None


def write_hello(hello: Hello) -> bytes:
    """Encode a greeting.

    The reference and the nonce are fixed width, so the greeting has no variable-length credential
    fields — there is no credential in it to be variable. The dictionary id trails them so a peer
    that stops reading there still parses: a missing id simply means no dictionary. The filter
    trails that in turn, length-prefixed with a ``u32`` because a filter naming two thousand
    accounts is far past 64 KiB.
    """
    out = bytearray(76)
    struct.pack_into("<HIBB", out, 0, PROTOCOL_VERSION, hello.streams, hello.codecs, hello.capabilities)
    out[8:40] = hello.key_ref
    out[40:72] = hello.client_nonce
    struct.pack_into("<I", out, 72, hello.dictionary_id)
    if hello.filter is not None:
        # Byte length, not character count: a filter is base58 today but the field is UTF-8.
        encoded = hello.filter.encode("utf-8")
        out += struct.pack("<I", len(encoded)) + encoded
    return bytes(out)


@dataclass(frozen=True, slots=True)
class Challenge:
    """The server's nonce and its proof that it holds your key.

    Sent before you have proved anything. Checking it is what tells you the peer is the node rather
    than something sitting in front of it.
    """

    server_nonce: bytes
    proof: bytes


def read_challenge(buf: bytes) -> Challenge:
    """Decode a challenge."""
    if len(buf) < 64:
        raise ProtocolError("challenge truncated")
    return Challenge(server_nonce=bytes(buf[0:32]), proof=bytes(buf[32:64]))


def write_prove(proof: bytes) -> bytes:
    """Encode your proof that you hold the key."""
    return bytes(proof)


@dataclass(frozen=True, slots=True)
class HelloAck:
    """What the server settled."""

    protocol_version: int
    session_id: int
    #: Streams actually granted — the intersection of what was asked and what the key permits.
    granted: int
    codec: Codec
    #: How often the server pings an idle connection.
    keepalive_ms: int
    #: Non-zero only when the server holds the same dictionary this client offered.
    dictionary_id: int


#: Bytes in an encoded acknowledgement.
HELLO_ACK_LEN = 22


def read_hello_ack(buf: bytes) -> HelloAck:
    """Decode an acknowledgement."""
    if len(buf) < HELLO_ACK_LEN:
        raise ProtocolError(f"hello ack needs {HELLO_ACK_LEN} bytes, have {len(buf)}")
    version, granted, session, codec = struct.unpack_from("<HIQB", buf, 0)
    # Byte 15 is reserved.
    keepalive, dictionary_id = struct.unpack_from("<HI", buf, 16)
    return HelloAck(
        protocol_version=version,
        granted=granted,
        session_id=session,
        codec=Codec(codec),
        keepalive_ms=keepalive,
        dictionary_id=dictionary_id,
    )


def read_error(buf: bytes) -> ServerError:
    """Decode an error message."""
    if len(buf) < 4:
        raise ProtocolError("error message truncated")
    code, length = struct.unpack_from("<HH", buf, 0)
    detail = bytes(buf[4 : 4 + length]).decode("utf-8", errors="replace")
    return ServerError(code, detail)


@dataclass(frozen=True, slots=True)
class Lag:
    """Frames this client missed, and where the stream resumes."""

    dropped: int
    resume_seq: int


def read_lag(buf: bytes) -> Lag:
    """Decode a lag report."""
    if len(buf) < 16:
        raise ProtocolError("lag message truncated")
    dropped, resume = struct.unpack_from("<QQ", buf, 0)
    return Lag(dropped=dropped, resume_seq=resume)


@dataclass(frozen=True, slots=True)
class FilterAck:
    """The server's verdict on a submitted filter."""

    accepted: bool
    cost: int
    detail: str


def read_filter_ack(buf: bytes) -> FilterAck:
    """Decode a filter acknowledgement."""
    if len(buf) < 8:
        raise ProtocolError("filter ack truncated")
    accepted = buf[0] != 0
    # Byte 1 is reserved.
    (cost,) = struct.unpack_from("<I", buf, 2)
    (length,) = struct.unpack_from("<H", buf, 6)
    detail = bytes(buf[8 : 8 + length]).decode("utf-8", errors="replace")
    return FilterAck(accepted=accepted, cost=cost, detail=detail)


@dataclass(frozen=True, slots=True)
class NamedFilter:
    """One filter, with the name the subscriber chose for it."""

    #: The filter itself.
    spec: dict
    #: What you call it. Comes back in a refusal so you know which one was at fault.
    name: str | None = None


def write_set_filter(filters: list[NamedFilter]) -> bytes:
    """Encode a filter submission. The frame length delimits it, so the JSON needs no prefix.

    The whole set replaces whatever the connection carried, and is accepted or refused together — a
    connection is never left holding part of what was asked for.
    """
    body = {
        "filters": [
            {"name": f.name, "spec": f.spec} if f.name is not None else {"spec": f.spec}
            for f in filters
        ]
    }
    return json.dumps(body, separators=(",", ":")).encode("utf-8")


def dictionary_id(dictionary: bytes) -> int:
    """The content hash a dictionary is identified by.

    FNV-1a folded to 32 bits, matching the server exactly. Two peers agree on a dictionary by
    agreeing on this number, so it must be computed the same way on both sides or the dictionary is
    silently never used.
    """
    if not dictionary:
        return 0
    hash_ = 0xCBF29CE484222325
    prime = 0x100000001B3
    mask = 0xFFFFFFFFFFFFFFFF
    for byte in dictionary:
        hash_ = ((hash_ ^ byte) * prime) & mask
    folded = (hash_ ^ (hash_ >> 32)) & 0xFFFFFFFF
    # Zero means "no dictionary", so a hash that lands there is nudged rather than misread.
    return 1 if folded == 0 else folded
